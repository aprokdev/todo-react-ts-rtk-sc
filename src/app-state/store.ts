import hideCompletedReducer from '@app-state/hideCompleted/hideCompletedSlice';
import isSavedTodosReducer from '@app-state/isSavedTodos/isSavedTodosSlice';
import sortingTitleReducer from '@app-state/sortingTitle/sortingTitleSlice';
import todosReducer from '@app-state/todos/todosSlice';
import { PreloadedState, combineReducers, configureStore } from '@reduxjs/toolkit';

export const rootReducer = combineReducers({
    todos: todosReducer,
    sortingTitle: sortingTitleReducer,
    hideCompleted: hideCompletedReducer,
    isSavedTodos: isSavedTodosReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
    const store = configureStore({
        reducer: rootReducer,
        devTools: true,
        preloadedState,
    });

    let isSavedTodos = true;
    // adds subscribtion on every action:
    store.subscribe(() => {
        // saves current todos and sorting states to LocalStorage only if action type is not "clearLocalStorage":
        isSavedTodos = store.getState().isSavedTodos;
        if (isSavedTodos) {
            localStorage.setItem('listTodos', JSON.stringify(store.getState().todos));
            localStorage.setItem('sortingTitle', JSON.stringify(store.getState().sortingTitle));
        }
        isSavedTodos = true;
    });

    return store;
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
