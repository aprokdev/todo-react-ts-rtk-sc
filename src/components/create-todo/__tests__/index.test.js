import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Provider } from 'react-redux';
import { act, create } from 'react-test-renderer';
import { setupStore } from '~app-state/store';
import { renderWithProvider } from '~src/app-state/test-utils';
import CreateTodo from '../index';

const store = setupStore();

function TestCreateTodo() {
    return (
        <Provider store={store}>
            <CreateTodo />
        </Provider>
    );
}

describe('CreateTodo', () => {
    test('matches snapshot', () => {
        let tree;
        act(() => {
            tree = create(<TestCreateTodo />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test("CreateTodo's main input and button are exist in document", async () => {
        expect(screen.queryByTestId('todo-input')).toBeNull();
        expect(screen.queryByTestId('todo-create-btn')).toBeNull();
        renderWithProvider(<CreateTodo />);
        expect(screen.queryByTestId('todo-input')).toBeInTheDocument();
        expect(screen.queryByTestId('todo-create-btn')).toBeInTheDocument();
    });

    test('input onChange handler updates value in CreateTodo', async () => {
        const user = userEvent.setup();
        renderWithProvider(<CreateTodo />);
        const todoInput = screen.getByTestId('todo-input');
        expect(todoInput.value).toBe('');
        await user.type(todoInput, 't');
        expect(todoInput.value).toBe('t');
        await user.type(todoInput, 'e');
        expect(todoInput.value).toBe('te');
    });

    test('"create todo" button changes disabled state if input value changes', async () => {
        const user = userEvent.setup();
        renderWithProvider(<CreateTodo />);
        const todoInput = screen.getByTestId('todo-input');
        const addTodoBtn = screen.getByTestId('todo-create-btn');
        expect(addTodoBtn).toBeDisabled();
        expect(todoInput.value).toBe('');
        await user.type(todoInput, 't');
        expect(todoInput.value).toBe('t');
        expect(addTodoBtn).not.toBeDisabled();
        await user.type(todoInput, 't');
        expect(todoInput.value).toBe('tt');
        expect(addTodoBtn).not.toBeDisabled();
        await user.type(todoInput, '{Backspace}{Backspace}');
        expect(todoInput.value).toBe('');
        expect(addTodoBtn).toBeDisabled();
    });
});
