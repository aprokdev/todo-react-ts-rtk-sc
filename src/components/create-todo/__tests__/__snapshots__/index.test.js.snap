// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTodo matches snapshot 1`] = `
<form
  className="create-todo"
>
  <input
    className="input"
    data-testid="todo-input"
    id="new-todo-input"
    onChange={[Function]}
    placeholder="Write new task here..."
    type="text"
    value=""
  />
  <button
    className="button"
    data-testid="todo-create-btn"
    disabled={true}
    onClick={[Function]}
    type="submit"
  >
    Add
  </button>
</form>
`;
